src/osmdata.cpp

```diff
@@ -34,21 +34,12 @@ osmdata_t::osmdata_t(std::unique_ptr<dependency_manager_t> dependency_manager,
     assert(!m_outs.empty());
 }
 
-/**
- * For modify and delete member functions a middle_t is not enough, an object
- * of the derived class slim_middle_t is needed. This function does the
- * conversion. It should always succeed, because the modify and delete
- * functions are never called for non-slim middles.
- */
-slim_middle_t &osmdata_t::slim_middle() const noexcept
-{
-    auto *slim = dynamic_cast<slim_middle_t *>(m_mid.get());
-    assert(slim);
-    return *slim;
-}
-
 void osmdata_t::node(osmium::Node const &node)
 {
+    if (!m_bbox.valid() || node.deleted() || m_bbox.contains(node.location())) {
```

Comment:
contains() is now called twice. Not the cheapest of functions.

More importantly: you either need to move the node.location().valid() check to the front of the function or make sure that mid can handle invalid locations. I'd be in favour of the first solution. (Bonus points for adding a test for input with invalid locations.)

Reply:
In an additional commit I have refactored this and added a test.